pallets/usage/src/lib.rs

```diff
@@ -0,0 +1,405 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[macro_use]
+extern crate alloc;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+// #[cfg(feature = "runtime-benchmarks")]
+// mod benchmarking;
+pub mod weights;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+mod offchain;
+
+pub type Authorship<T> = pallet_authorship::Pallet<T>;
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::offchain::PayeeInfo;
+    use super::*;
+    use common::{
+        value_tree::ValueView, GasToFeeConverter, Message, PaymentProvider, GAS_VALUE_PREFIX,
+    };
+    use frame_support::{
+        dispatch::{DispatchError, DispatchResultWithPostInfo},
+        pallet_prelude::*,
+        traits::{Currency, Get, ReservableCurrency},
+    };
+    use frame_system::{offchain::SendTransactionTypes, pallet_prelude::*, RawOrigin};
+    use primitive_types::H256;
+    use sp_core::offchain::Duration;
+    use sp_runtime::{
+        offchain::{
+            storage::StorageValueRef,
+            storage_lock::{StorageLock, Time},
+        },
+        traits::{SaturatedConversion, Saturating},
+        Perbill,
+    };
+    use sp_std::prelude::*;
+
+    #[pallet::config]
+    pub trait Config:
+        frame_system::Config + pallet_authorship::Config + SendTransactionTypes<Call<Self>>
+    {
+        /// Because this pallet emits events, it depends on the runtime's definition of an event.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+        /// Gas and value transfer currency
+        type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+
+        /// Gas to Currency converter
+        type GasConverter: GasToFeeConverter<Balance = BalanceOf<Self>>;
+
+        /// Type providing interface for making payment in currency units
+        type PaymentProvider: PaymentProvider<Self::AccountId, Balance = BalanceOf<Self>>;
```

Comment:
The reason for having this associated type as opposed to using the `Currency` type's `repatriate_reserved` method directly is that we don't want to impose additional trait bound on the `T::AccountId` type (to implement the `common::Origin` trait) in order to avoid conflict, created by the pallet's `Config` trait inheriting from the `SendTransactionTypes<Call<Self>>` because the type parameter `Call<Self>` is then subject to the same trait bound `T::AccountId: common::Origin` which is not the case for the `SendTransactionTypes<T>`.